color_refinement.py

```diff
@@ -86,7 +88,86 @@ def color_refine(coloring: Coloring) -> Coloring:
+    """
+    # Start with first color
+    qlist = DoubleLinkedList()
+    for c in sorted(coloring.colors):
+        qlist.append(c)
+    debug('Queue', qlist)
+
+    while(len(qlist) > 0):
+        # Start refining with the first color from the queue.
+        current_color = qlist.pop_left()
+        counter = generate_neighbour_count_with_color(graph, current_color)
+
+        for color_class in counter.keys():
+            # Will loop over all the colors in the graph and refine them.
+            debug('Refining the following color:', color_class)
+            neighbour_map = counter[color_class]
+            # Partitions class 'c' into cells according to #neighbours of current_color
+            vertices_of_c = list(neighbour_map.keys())
+            debug('Vertices of color', color_class, vertices_of_c)
+            debug('Neighbours', neighbour_map)
+            # Keep split_count so the first vertices you encounter are not recolored
+            split_count = 0
+            # Keep a list of the color_classes from this loop so we can count them after the loop and see which one
+            # is the smallest and should be added to the queue
+            new_color_classes = []
+
+            while len(vertices_of_c) > 0:
+                u = vertices_of_c.pop()
+                new_color = color_class
+                if split_count > 0:
+                    new_color = coloring.next_color()
+                    coloring.recolor(u, new_color, u.colornum)
+
+                for v in list(vertices_of_c):
+                    n_neighbours_u = neighbour_map[u]
+                    n_neighbours_v = neighbour_map[v]
+                    # Compare the amount of neighbours of u with the amount of neighbours of v
+                    # If they are equal u and v are in the same cell
+                    # If the split count is larger then zero they should both be colored with the same color
+                    if n_neighbours_u == n_neighbours_v:
+                        if split_count > 0:
+                            coloring.recolor(v, new_color, v.colornum)
+                        vertices_of_c.remove(v)
+                split_count += 1
+                # Each color_class is added to the list
+                new_color_classes.append(new_color)
+            # The smallest_color here is the first color added to the list, which is the original color
+            smallest_color = new_color_classes[0]
+            if split_count > 1:
+                debug('New color classes:', new_color_classes)
+                # If the original color is in the queue, all the other colors should be added to the queue
+                if qlist.find(smallest_color) is not None:
+                    for color in new_color_classes:
+                        if qlist.find(color) is None:
+                            qlist.append(color)
+                # Otherwise the color with the least amount over vertices should be added to the queue
```

Comment:
Als de lijst > 2, moet je dan lengte -1 aantal Colors toevoegen? Of alleen maar de kleinste?

Reply:
Dat weet ik niet :) Aangezien alle voorbeelden met het splitsen in twee waren hebben we dit aangehouden, maar misschien heb je wel gelijk :)